File: part1/QuickUnionWeight/QuickUnionWeight.java
import java.util.Arrays;

public class QuickUnionWeight {

    private int[] id;
    private int[] sz;

    public QuickUnionWeight(int N) {
        id = new int[N];
        sz = new int[N];
        for (int i = 0; i < N; i++) {
            id[i] = i;
            sz[i] = 1;
        }
    }

    // 回溯到最上层的父节点
    private int root(int i) {
        while (i != id[i]) {
            i = id[i];
        }
        return i;
    }

    // 判断是否有同样的最上层父节点
    public boolean connected(int p, int q) {
        return root(p) == root(q);
    }

    public void union(int p, int q) {
        int i = root(p); // 6
        int j = root(q); // 0
        if (i == j) {
            return;
        }
        if (sz[i] < sz[j]) { // 2 1
            id[i] = j;
            sz[j] += sz[i];
        } else {
            id[j] = i;
            sz[i] += sz[j];
        }
    }

    public static void main(String[] args) {
        QuickUnionWeight quw = new QuickUnionWeight(10);
        quw.union(4, 3);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(3, 8);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(6, 5);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(9, 4);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(2, 1);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(5, 0);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(7, 2);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        quw.union(6, 1);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");
        System.out.println(quw.connected(7, 3));
        quw.union(7, 3);
        System.out.println(Arrays.toString(quw.id));
        System.out.println(Arrays.toString(quw.sz));
        System.out.println("--------");

    }
}

File: part1/QuickUnionWeight/QuickUnionWeight.js
class QuickUnionWeight {
  id = [];
  sz = [];

  constructor(N) {
    this.id = new Array(N);
    this.sz = new Array(N);
    for (let i = 0; i < N; i++) {
      this.id[i] = i;
      this.sz[i] = 1;
    }
  }

  root(i) {
    while (i !== this.id[i]) {
      i = this.id[i];
    }
    return i;
  }

  connected(p, q) {
    return this.root(p) === this.root(q);
  }

  union(p, q) {
    const i = this.root(p);
    const j = this.root(q);

    if (i === j) {
      return;
    }

    if (this.sz[i] < this.sz[j]) {
      this.id[i] = j;
      this.sz[j] += this.sz[i];
    } else {
      this.id[j] = i;
      this.sz[i] += this.sz[j];
    }
  }
}

const quw = new QuickUnionWeight(10);
quw.union(4, 3);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(3, 8);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(6, 5);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(9, 4);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(2, 1);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(5, 0);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(7, 2);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
quw.union(6, 1);
quw.union(7, 3);
console.info(quw.id);
console.info(quw.sz);
console.info("--------");
